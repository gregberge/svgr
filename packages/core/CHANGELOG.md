# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [4.3.0](https://github.com/smooth-code/svgr/tree/master/packages/core/compare/v4.2.0...v4.3.0) (2019-05-28)

**Note:** Version bump only for package @svgr/core





# [4.2.0](https://github.com/smooth-code/svgr/tree/master/packages/core/compare/v4.1.0...v4.2.0) (2019-04-11)


### Bug Fixes

* keep viewBox when dimensions are removed ([#281](https://github.com/smooth-code/svgr/tree/master/packages/core/issues/281)) ([f476c8e](https://github.com/smooth-code/svgr/tree/master/packages/core/commit/f476c8e))


### Features

* add expo option ([#289](https://github.com/smooth-code/svgr/tree/master/packages/core/issues/289)) ([978db3e](https://github.com/smooth-code/svgr/tree/master/packages/core/commit/978db3e))





# [4.1.0](https://github.com/smooth-code/svgr/compare/v4.0.4...v4.1.0) (2018-11-24)

**Note:** Version bump only for package @svgr/core





## [4.0.3](https://github.com/smooth-code/svgr/compare/v4.0.2...v4.0.3) (2018-11-13)


### Bug Fixes

* upgrade dependencies ([7e2195f](https://github.com/smooth-code/svgr/commit/7e2195f))





## [4.0.2](https://github.com/smooth-code/svgr/compare/v4.0.1...v4.0.2) (2018-11-08)

**Note:** Version bump only for package @svgr/core





## [4.0.1](https://github.com/smooth-code/svgr/compare/v4.0.0...v4.0.1) (2018-11-08)

**Note:** Version bump only for package @svgr/core





# [4.0.0](https://github.com/smooth-code/svgr/compare/v3.1.0...v4.0.0) (2018-11-04)


### Features

* **svgo:** prefix ids by default ([06c338d](https://github.com/smooth-code/svgr/commit/06c338d)), closes [#210](https://github.com/smooth-code/svgr/issues/210)
* **v4:** new architecture ([ac8b8ca](https://github.com/smooth-code/svgr/commit/ac8b8ca))
* allow dynamic properties in replaceAttrValues option ([15f55fe](https://github.com/smooth-code/svgr/commit/15f55fe)), closes [#205](https://github.com/smooth-code/svgr/issues/205)


### BREAKING CHANGES

* **v4:** - `template` option must now returns a Babel AST
- `@svgr/core` does not include svgo & prettier by default





# [3.1.0](https://github.com/smooth-code/svgr/compare/v3.0.0...v3.1.0) (2018-10-05)


### Bug Fixes

* style & custom SVG properties ([#203](https://github.com/smooth-code/svgr/issues/203)) ([f8b2212](https://github.com/smooth-code/svgr/commit/f8b2212)), closes [#199](https://github.com/smooth-code/svgr/issues/199) [#201](https://github.com/smooth-code/svgr/issues/201)


### Features

* allow Mask & Image on React Native ([#202](https://github.com/smooth-code/svgr/issues/202)) ([0256bc0](https://github.com/smooth-code/svgr/commit/0256bc0))





<a name="3.0.0"></a>
# [3.0.0](https://github.com/smooth-code/svgr/compare/v2.4.1...v3.0.0) (2018-10-01)


### Bug Fixes

* fix --icon + --no-dimensions ([7535693](https://github.com/smooth-code/svgr/commit/7535693)), closes [#141](https://github.com/smooth-code/svgr/issues/141)
* fix expandProps when position is not allowed ([45522fc](https://github.com/smooth-code/svgr/commit/45522fc))


### Features

* **config:** improve runtime config ([e52cdce](https://github.com/smooth-code/svgr/commit/e52cdce)), closes [#192](https://github.com/smooth-code/svgr/issues/192)
* **template:** expose `getProps` util for template ([5cb238e](https://github.com/smooth-code/svgr/commit/5cb238e)), closes [#187](https://github.com/smooth-code/svgr/issues/187)
* add synchronous API ([169eb2f](https://github.com/smooth-code/svgr/commit/169eb2f)), closes [#185](https://github.com/smooth-code/svgr/issues/185)
* always prefix component name with "Svg" ([f71aa7a](https://github.com/smooth-code/svgr/commit/f71aa7a)), closes [#190](https://github.com/smooth-code/svgr/issues/190)
* do not remove style tag ([a4ce09a](https://github.com/smooth-code/svgr/commit/a4ce09a)), closes [#191](https://github.com/smooth-code/svgr/issues/191)
* new "expandProps" option ([bb95828](https://github.com/smooth-code/svgr/commit/bb95828)), closes [#170](https://github.com/smooth-code/svgr/issues/170)
* remove "svgAttributes" option ([4e46a5d](https://github.com/smooth-code/svgr/commit/4e46a5d)), closes [#173](https://github.com/smooth-code/svgr/issues/173)
* use forwardRef on React Native ([4bdd989](https://github.com/smooth-code/svgr/commit/4bdd989)), closes [#184](https://github.com/smooth-code/svgr/issues/184)
* use React.forwardRef ([cbee51c](https://github.com/smooth-code/svgr/commit/cbee51c)), closes [#184](https://github.com/smooth-code/svgr/issues/184)


### BREAKING CHANGES

* "--no-expand-props" is now replaced by "--expand-props none". You can now specify a position "start" or "end" for "expandProps"
property.
* `svgAttributes` has been removed, please use `svgProps` instead.
* "ref" option now uses `React.forwardRef`. You don't have to use "svgRef"
prop, just use "ref" and it will work. `React.forwardRef` requires React
> 16.3.
* Style tag will no longer be automatically removed. SVGO should handle it
correctly using "inlineStyles" plugin. If you want to remove them,
enable "removeStyleElement" plugin in your SVGO config.
* **config:** - Runtime configuration is always loaded (even with Node API `convert`)
- In CLI, "--config" is now "--config-file"; this new option can be used
everywhere





<a name="2.4.1"></a>
## [2.4.1](https://github.com/smooth-code/svgr/compare/v2.4.0...v2.4.1) (2018-09-16)


### Bug Fixes

* **config:** fix custom config & default options ([#176](https://github.com/smooth-code/svgr/issues/176)) ([9a6c40b](https://github.com/smooth-code/svgr/commit/9a6c40b))





<a name="2.4.0"></a>
# [2.4.0](https://github.com/smooth-code/svgr/compare/v2.3.0...v2.4.0) (2018-09-16)


### Bug Fixes

* use literal instead of litteral ([7849fd4](https://github.com/smooth-code/svgr/commit/7849fd4))


### Features

* allow to spread props at the start ([#166](https://github.com/smooth-code/svgr/issues/166)) ([cd659dc](https://github.com/smooth-code/svgr/commit/cd659dc))
* **upgrade:** h2x@1.1.0 (jsdom@12.0.0) & others ([2d9b7bd](https://github.com/smooth-code/svgr/commit/2d9b7bd))
* new option "svgProps" ([#172](https://github.com/smooth-code/svgr/issues/172)) ([9657110](https://github.com/smooth-code/svgr/commit/9657110))





<a name="2.2.0"></a>
# [2.2.0](https://github.com/smooth-code/svgr/compare/v2.1.1...v2.2.0) (2018-08-13)


### Bug Fixes

* remove null-byte characters ([#154](https://github.com/smooth-code/svgr/issues/154)) ([de7f8a7](https://github.com/smooth-code/svgr/commit/de7f8a7)), closes [#153](https://github.com/smooth-code/svgr/issues/153)


### Features

* **core:** pass info to SVGO ([2b2353b](https://github.com/smooth-code/svgr/commit/2b2353b)), closes [#152](https://github.com/smooth-code/svgr/issues/152)





<a name="2.1.1"></a>
## [2.1.1](https://github.com/smooth-code/svgr/compare/v2.1.0...v2.1.1) (2018-07-11)


### Bug Fixes

* **core:** config conflict with icon option ([#137](https://github.com/smooth-code/svgr/issues/137)) ([e13a99a](https://github.com/smooth-code/svgr/commit/e13a99a))




<a name="2.1.0"></a>
# [2.1.0](https://github.com/smooth-code/svgr/compare/v2.0.0...v2.1.0) (2018-07-08)


### Features

* add .editorconfig support ([#129](https://github.com/smooth-code/svgr/issues/129)) ([968fd82](https://github.com/smooth-code/svgr/commit/968fd82))
* **cli:** support custom filename cases ([#136](https://github.com/smooth-code/svgr/issues/136)) ([4922f7a](https://github.com/smooth-code/svgr/commit/4922f7a)), closes [#118](https://github.com/smooth-code/svgr/issues/118)
