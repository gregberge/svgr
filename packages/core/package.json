{
  "name": "@svgr/core",
  "description": "Transform SVG into React Components.",
  "version": "3.1.0",
  "main": "lib/index.js",
  "repository": "git@github.com:smooth-code/svgr.git",
  "author": "Greg Berg√© <berge.greg@gmail.com>",
  "keywords": [
    "svgr",
    "svg",
    "react",
    "core",
    "api"
  ],
  "engines": {
    "node": ">=8"
  },
  "license": "MIT",
  "scripts": {
    "prebuild": "rm -rf lib/",
    "build": "babel --config-file ../../babel.config.js -d lib --ignore \"**/*.test.js\" src",
    "prepublishOnly": "yarn run build"
  },
  "dependencies": {
    "@babel/core": "^7.1.2",
    "@babel/generator": "^7.1.3",
    "camelcase": "^5.0.0",
    "cosmiconfig": "^5.0.6",
    "h2x-core": "^1.1.1",
    "h2x-plugin-jsx": "^1.2.0",
    "hast-util-to-babel-ast": "^3.1.0",
    "merge-deep": "^3.0.2",
    "prettier": "^1.14.3",
    "rehype-parse": "^5.0.0",
    "svgo": "^1.1.1",
    "unified": "^7.0.0",
    "vfile": "^3.0.0"
  }
}
