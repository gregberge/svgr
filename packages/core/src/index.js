/* eslint-disable-next-line import/no-cycle */
export { default } from './convert'
export * from './config'
export * from './util'

// H2X plugins
export { default as emSize } from './h2x/emSize'
export { default as expandProps } from './h2x/expandProps'
export { default as removeComments } from './h2x/removeComments'
export { default as removeDimensions } from './h2x/removeDimensions'
export { default as removeStyle } from './h2x/removeStyle'
export { default as replaceAttrValues } from './h2x/replaceAttrValues'
export { default as stripAttribute } from './h2x/stripAttribute'
export { default as svgAttributes } from './h2x/svgAttributes'
export { default as svgProps } from './h2x/svgProps'
export { default as svgRef } from './h2x/svgRef'
export { default as titleProp } from './h2x/titleProp'
export { default as toReactNative } from './h2x/toReactNative'

// Templates
export { default as reactDomTemplate } from './templates/reactDomTemplate'
export { default as reactNativeTemplate } from './templates/reactNativeTemplate'
