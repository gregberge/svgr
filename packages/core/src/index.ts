export { transform } from './transform'
export * from './config'
export type { State } from './state'
export type { Plugin, ConfigPlugin } from './plugins'
