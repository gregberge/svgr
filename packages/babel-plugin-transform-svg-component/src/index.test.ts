import { transform } from '@babel/core'
import plugin, { Options } from '.'

const defaultOptions = {
  namedExport: 'ReactComponent',
  state: { componentName: 'SvgComponent' },
}

const testPlugin =
  (language: string) =>
  (code: string, options: Partial<Options> = {}) => {
    const result = transform(code, {
      plugins: [
        '@babel/plugin-syntax-jsx',
        [
          plugin,
          {
            typescript: language === 'typescript',
            ...defaultOptions,
            ...options,
          },
        ],
      ],
      configFile: false,
    })

    if (!result) {
      throw new Error(`No result`)
    }

    return result
  }

describe('plugin', () => {
  describe.each(['javascript', 'typescript'])('%s', (language) => {
    it('transforms whole program', () => {
      const { code } = testPlugin(language)('<svg><g /></svg>')
      expect(code).toMatchSnapshot()
    })

    describe('with "native" option', () => {
      it('adds import from "react-native-svg"', () => {
        const { code } = testPlugin(language)('<Svg><g /></Svg>', {
          native: true,
        })
        expect(code).toMatchSnapshot()
      })
    })

    describe('with "ref" option', () => {
      it('adds ForwardRef component', () => {
        const { code } = testPlugin(language)('<svg><g /></svg>', {
          ref: true,
        })
        expect(code).toMatchSnapshot()
      })
    })

    describe('with "titleProp"', () => {
      it('adds "titleProp" and "titleId" prop', () => {
        const { code } = testPlugin(language)('<svg><g /></svg>', {
          titleProp: true,
        })
        expect(code).toMatchSnapshot()
      })
    })

    describe('with "titleProp" and "expandProps"', () => {
      it('adds "titleProp", "titleId" props and expands props', () => {
        const { code } = testPlugin(language)('<svg><g /></svg>', {
          ...defaultOptions,
          expandProps: true,
          titleProp: true,
        })
        expect(code).toMatchSnapshot()
      })
    })

    describe('with "expandProps"', () => {
      it('add props', () => {
        const { code } = testPlugin(language)('<svg><g /></svg>', {
          ...defaultOptions,
          state: { componentName: 'SvgComponent' },
          expandProps: true,
        })
        expect(code).toMatchSnapshot()
      })
    })

    describe('with "ref" and "expandProps" option', () => {
      it('expands props', () => {
        const { code } = testPlugin(language)('<svg><g /></svg>', {
          state: { componentName: 'SvgComponent' },
          expandProps: true,
          ref: true,
        })
        expect(code).toMatchSnapshot()
      })
    })

    describe('with "native", "ref" option', () => {
      it('adds import from "react-native-svg" and adds ForwardRef component', () => {
        const { code } = testPlugin(language)('<Svg><g /></Svg>', {
          state: { componentName: 'SvgComponent' },
          native: true,
          ref: true,
        })
        expect(code).toMatchSnapshot()
      })
    })

    describe('with "native" and "expandProps" option', () => {
      it('adds import from "react-native-svg" and adds props', () => {
        const { code } = testPlugin(language)('<Svg><g /></Svg>', {
          state: { componentName: 'SvgComponent' },
          native: true,
          expandProps: true,
        })
        expect(code).toMatchSnapshot()
      })
    })

    describe('with "native", "ref" and "expandProps" option', () => {
      it('adds import from "react-native-svg" and adds props and adds ForwardRef component', () => {
        const { code } = testPlugin(language)('<Svg><g /></Svg>', {
          state: { componentName: 'SvgComponent' },
          native: true,
          expandProps: true,
          ref: true,
        })
        expect(code).toMatchSnapshot()
      })
    })

    describe('with "memo" option', () => {
      it('wrap component in "React.memo"', () => {
        const { code } = testPlugin(language)('<svg><g /></svg>', {
          state: { componentName: 'SvgComponent' },
          memo: true,
        })
        expect(code).toMatchSnapshot()
      })
    })

    describe('with both "memo" and "ref" option', () => {
      it('wrap component in "React.memo" and "React.forwardRef"', () => {
        const { code } = testPlugin(language)('<svg><g /></svg>', {
          state: { componentName: 'SvgComponent' },
          memo: true,
          ref: true,
        })
        expect(code).toMatchSnapshot()
      })
    })

    describe('with "namedExport" option and "previousExport" state', () => {
      it('has custom named export', () => {
        const { code } = testPlugin(language)('<svg><g /></svg>', {
          state: {
            componentName: 'SvgComponent',
            caller: {
              previousExport: `var img = new Image(); img.src = '...'; export default img;`,
            },
          },
          namedExport: 'Component',
        })
        expect(code).toMatchSnapshot()
      })
    })

    describe('with "namedExport" and "exportType" option and without "previousExport" state', () => {
      it('exports via named export', () => {
        const { code } = testPlugin(language)('<svg><g /></svg>', {
          state: {
            componentName: 'SvgComponent',
            caller: { previousExport: null },
          },
          namedExport: 'ReactComponent',
          exportType: 'named',
        })
        expect(code).toMatchSnapshot()
      })
    })

    describe('custom templates', () => {
      it('support basic template', () => {
        const { code } = testPlugin(language)('<svg><g /></svg>', {
          template: ({ jsx }, { tpl }) => tpl`import * as React from 'react';
          const MyComponent = () => ${jsx}
          export default MyComponent
        `,
          state: { componentName: 'SvgComponent' },
        })
        expect(code).toMatchSnapshot()
      })

      it('supports JSX template', () => {
        const { code } = testPlugin(language)('<svg><g /></svg>', {
          template: ({ jsx }, { tpl }) => {
            return tpl`import * as React from 'react';
            const MyComponent = () => <main>{${jsx}}</main>
            export default MyComponent
          `
          },
          state: { componentName: 'SvgComponent' },
        })
        expect(code).toMatchSnapshot()
      })

      it('supports TypeScript template', () => {
        const { code } = testPlugin(language)('<svg><g /></svg>', {
          template: ({ jsx }, { tpl }) => {
            return tpl`
              import * as React from 'react';
              const MyComponent = (props: React.SVGProps<SVGSVGElement>) => ${jsx};
              export default MyComponent;
            `
          },
          typescript: true,
          state: { componentName: 'SvgComponent' },
        })
        expect(code).toMatchSnapshot()
      })

      it('supports template that does not return an array', () => {
        const { code } = testPlugin(language)('<svg><g /></svg>', {
          template: ({ jsx }, { tpl }) => tpl`${jsx}`,
          state: { componentName: 'SvgComponent' },
        })
        expect(code).toMatchSnapshot()
      })

      it('supports type annotation on component', () => {
        const { code } = testPlugin(language)('<svg><g /></svg>', {
          typescript: true,
          template: (
            { jsx, imports, interfaces, componentName, exports },
            { tpl },
          ) => tpl`
          ${imports}
          ${interfaces}
          interface Props { x?: string }
          export const ${`${componentName}:React.FC<Props>`} = ({ x }) => {
            return (${jsx});
          }
          ${exports}`,
          state: { componentName: 'SvgComponent' },
        })
        expect(code).toMatchSnapshot()
      })

      it('supports comments in templates', () => {
        const { code } = testPlugin(language)('<svg><g /></svg>', {
          template: ({ jsx }, { tpl }) => tpl`
          /**
           * Comment
           */
          const MyComponent = () => ${jsx}

          export default MyComponent;
          `,
          state: { componentName: 'SvgComponent' },
        })
        expect(code).toMatchSnapshot()
      })
    })

    describe('#jsxRuntime', () => {
      it('supports "automatic" jsxRuntime', () => {
        const { code } = testPlugin(language)('<svg><g /></svg>', {
          jsxRuntime: 'automatic',
        })
        expect(code).toMatchSnapshot()
      })

      it('supports "classic" jsxRuntime', () => {
        const { code } = testPlugin(language)('<svg><g /></svg>', {
          jsxRuntime: 'classic',
        })
        expect(code).toMatchSnapshot()
      })

      it('allows to specify a custom "classic" jsxRuntime using "specifiers"', () => {
        const { code } = testPlugin(language)('<svg><g /></svg>', {
          jsxRuntime: 'classic',
          jsxRuntimeImport: { specifiers: ['h'], source: 'preact' },
        })
        expect(code).toMatchSnapshot()
      })

      it('allows to specify a custom "classic" jsxRuntime using "namespace"', () => {
        const { code } = testPlugin(language)('<svg><g /></svg>', {
          jsxRuntime: 'classic',
          jsxRuntimeImport: { namespace: 'Preact', source: 'preact' },
        })
        expect(code).toMatchSnapshot()
      })

      it('throws with invalid configuration', () => {
        expect(() => {
          testPlugin(language)('<svg><g /></svg>', {
            jsxRuntime: 'classic',
            jsxRuntimeImport: { source: 'preact' },
          })
        }).toThrow(
          'Specify either "namespace" or "specifiers" in "jsxRuntimeImport" option',
        )
      })
    })

    it('allows to specify a different import source', () => {
      const { code } = testPlugin(language)('<svg><g /></svg>', {
        memo: true,
        ref: true,
        importSource: 'preact/compat',
        jsxRuntimeImport: { specifiers: ['h'], source: 'preact' },
      })
      expect(code).toMatchSnapshot()
    })
  })
})
