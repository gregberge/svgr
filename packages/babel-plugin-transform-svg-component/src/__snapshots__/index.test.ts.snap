// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugin javascript #jsxRuntime allows to specify a custom "classic" jsxRuntime using "namespace" 1`] = `
"import * as Preact from \\"preact\\";

const SvgComponent = () => <svg><g /></svg>;

export default SvgComponent;"
`;

exports[`plugin javascript #jsxRuntime allows to specify a custom "classic" jsxRuntime using "specifiers" 1`] = `
"import { h } from \\"preact\\";

const SvgComponent = () => <svg><g /></svg>;

export default SvgComponent;"
`;

exports[`plugin javascript #jsxRuntime supports "automatic" jsxRuntime 1`] = `
"const SvgComponent = () => <svg><g /></svg>;

export default SvgComponent;"
`;

exports[`plugin javascript #jsxRuntime supports "classic" jsxRuntime 1`] = `
"import * as React from \\"react\\";

const SvgComponent = () => <svg><g /></svg>;

export default SvgComponent;"
`;

exports[`plugin javascript allows to specify a different import source 1`] = `
"import { h } from \\"preact\\";
import { forwardRef, memo } from \\"preact/compat\\";

const SvgComponent = (_, ref) => <svg><g /></svg>;

const ForwardRef = forwardRef(SvgComponent);
const Memo = memo(ForwardRef);
export default Memo;"
`;

exports[`plugin javascript custom templates support basic template 1`] = `
"import * as React from 'react';

const MyComponent = () => <svg><g /></svg>;

export default MyComponent;"
`;

exports[`plugin javascript custom templates supports JSX template 1`] = `
"import * as React from 'react';

const MyComponent = () => <main>{<svg><g /></svg>}</main>;

export default MyComponent;"
`;

exports[`plugin javascript custom templates supports TypeScript template 1`] = `
"import * as React from 'react';

const MyComponent = (props: React.SVGProps<SVGSVGElement>) => <svg><g /></svg>;

export default MyComponent;"
`;

exports[`plugin javascript custom templates supports comments in templates 1`] = `
"/**
 * Comment
 */
const MyComponent = () => <svg><g /></svg>;

export default MyComponent;"
`;

exports[`plugin javascript custom templates supports template that does not return an array 1`] = `"<svg><g /></svg>;"`;

exports[`plugin javascript custom templates supports type annotation on component 1`] = `
"import * as React from \\"react\\";
interface Props {
  x?: string;
}
export const SvgComponent:React.FC<Props> = ({
  x
}) => {
  return <svg><g /></svg>;
};
export default SvgComponent;"
`;

exports[`plugin javascript transforms whole program 1`] = `
"import * as React from \\"react\\";

const SvgComponent = () => <svg><g /></svg>;

export default SvgComponent;"
`;

exports[`plugin javascript with "expandProps" add props 1`] = `
"import * as React from \\"react\\";

const SvgComponent = props => <svg><g /></svg>;

export default SvgComponent;"
`;

exports[`plugin javascript with "memo" option wrap component in "React.memo" 1`] = `
"import * as React from \\"react\\";
import { memo } from \\"react\\";

const SvgComponent = () => <svg><g /></svg>;

const Memo = memo(SvgComponent);
export default Memo;"
`;

exports[`plugin javascript with "namedExport" and "exportType" option and without "previousExport" state exports via named export 1`] = `
"import * as React from \\"react\\";

const SvgComponent = () => <svg><g /></svg>;

export { SvgComponent as ReactComponent };"
`;

exports[`plugin javascript with "namedExport" option and "previousExport" state has custom named export 1`] = `
"import * as React from \\"react\\";

const SvgComponent = () => <svg><g /></svg>;

export { SvgComponent as Component };
var img = new Image();
img.src = '...';
export default img;"
`;

exports[`plugin javascript with "native" and "expandProps" option adds import from "react-native-svg" and adds props 1`] = `
"import * as React from \\"react\\";
import Svg from \\"react-native-svg\\";

const SvgComponent = props => <Svg><g /></Svg>;

export default SvgComponent;"
`;

exports[`plugin javascript with "native" option adds import from "react-native-svg" 1`] = `
"import * as React from \\"react\\";
import Svg from \\"react-native-svg\\";

const SvgComponent = () => <Svg><g /></Svg>;

export default SvgComponent;"
`;

exports[`plugin javascript with "native", "ref" and "expandProps" option adds import from "react-native-svg" and adds props and adds ForwardRef component 1`] = `
"import * as React from \\"react\\";
import Svg from \\"react-native-svg\\";
import { forwardRef } from \\"react\\";

const SvgComponent = (props, ref) => <Svg><g /></Svg>;

const ForwardRef = forwardRef(SvgComponent);
export default ForwardRef;"
`;

exports[`plugin javascript with "native", "ref" option adds import from "react-native-svg" and adds ForwardRef component 1`] = `
"import * as React from \\"react\\";
import Svg from \\"react-native-svg\\";
import { forwardRef } from \\"react\\";

const SvgComponent = (_, ref) => <Svg><g /></Svg>;

const ForwardRef = forwardRef(SvgComponent);
export default ForwardRef;"
`;

exports[`plugin javascript with "ref" and "expandProps" option expands props 1`] = `
"import * as React from \\"react\\";
import { forwardRef } from \\"react\\";

const SvgComponent = (props, ref) => <svg><g /></svg>;

const ForwardRef = forwardRef(SvgComponent);
export default ForwardRef;"
`;

exports[`plugin javascript with "ref" option adds ForwardRef component 1`] = `
"import * as React from \\"react\\";
import { forwardRef } from \\"react\\";

const SvgComponent = (_, ref) => <svg><g /></svg>;

const ForwardRef = forwardRef(SvgComponent);
export default ForwardRef;"
`;

exports[`plugin javascript with "titleProp" adds "titleProp" and "titleId" prop 1`] = `
"import * as React from \\"react\\";

const SvgComponent = ({
  title,
  titleId
}) => <svg><g /></svg>;

export default SvgComponent;"
`;

exports[`plugin javascript with "titleProp" and "expandProps" adds "titleProp", "titleId" props and expands props 1`] = `
"import * as React from \\"react\\";

const SvgComponent = ({
  title,
  titleId,
  ...props
}) => <svg><g /></svg>;

export default SvgComponent;"
`;

exports[`plugin javascript with both "memo" and "ref" option wrap component in "React.memo" and "React.forwardRef" 1`] = `
"import * as React from \\"react\\";
import { forwardRef, memo } from \\"react\\";

const SvgComponent = (_, ref) => <svg><g /></svg>;

const ForwardRef = forwardRef(SvgComponent);
const Memo = memo(ForwardRef);
export default Memo;"
`;

exports[`plugin typescript #jsxRuntime allows to specify a custom "classic" jsxRuntime using "namespace" 1`] = `
"import * as Preact from \\"preact\\";

const SvgComponent = () => <svg><g /></svg>;

export default SvgComponent;"
`;

exports[`plugin typescript #jsxRuntime allows to specify a custom "classic" jsxRuntime using "specifiers" 1`] = `
"import { h } from \\"preact\\";

const SvgComponent = () => <svg><g /></svg>;

export default SvgComponent;"
`;

exports[`plugin typescript #jsxRuntime supports "automatic" jsxRuntime 1`] = `
"const SvgComponent = () => <svg><g /></svg>;

export default SvgComponent;"
`;

exports[`plugin typescript #jsxRuntime supports "classic" jsxRuntime 1`] = `
"import * as React from \\"react\\";

const SvgComponent = () => <svg><g /></svg>;

export default SvgComponent;"
`;

exports[`plugin typescript allows to specify a different import source 1`] = `
"import { h } from \\"preact\\";
import { Ref, forwardRef, memo } from \\"preact/compat\\";

const SvgComponent = (_, ref: Ref<SVGSVGElement>) => <svg><g /></svg>;

const ForwardRef = forwardRef(SvgComponent);
const Memo = memo(ForwardRef);
export default Memo;"
`;

exports[`plugin typescript custom templates support basic template 1`] = `
"import * as React from 'react';

const MyComponent = () => <svg><g /></svg>;

export default MyComponent;"
`;

exports[`plugin typescript custom templates supports JSX template 1`] = `
"import * as React from 'react';

const MyComponent = () => <main>{<svg><g /></svg>}</main>;

export default MyComponent;"
`;

exports[`plugin typescript custom templates supports TypeScript template 1`] = `
"import * as React from 'react';

const MyComponent = (props: React.SVGProps<SVGSVGElement>) => <svg><g /></svg>;

export default MyComponent;"
`;

exports[`plugin typescript custom templates supports comments in templates 1`] = `
"/**
 * Comment
 */
const MyComponent = () => <svg><g /></svg>;

export default MyComponent;"
`;

exports[`plugin typescript custom templates supports template that does not return an array 1`] = `"<svg><g /></svg>;"`;

exports[`plugin typescript custom templates supports type annotation on component 1`] = `
"import * as React from \\"react\\";
interface Props {
  x?: string;
}
export const SvgComponent:React.FC<Props> = ({
  x
}) => {
  return <svg><g /></svg>;
};
export default SvgComponent;"
`;

exports[`plugin typescript transforms whole program 1`] = `
"import * as React from \\"react\\";

const SvgComponent = () => <svg><g /></svg>;

export default SvgComponent;"
`;

exports[`plugin typescript with "expandProps" add props 1`] = `
"import * as React from \\"react\\";
import { SVGProps } from \\"react\\";

const SvgComponent = (props: SVGProps<SVGSVGElement>) => <svg><g /></svg>;

export default SvgComponent;"
`;

exports[`plugin typescript with "memo" option wrap component in "React.memo" 1`] = `
"import * as React from \\"react\\";
import { memo } from \\"react\\";

const SvgComponent = () => <svg><g /></svg>;

const Memo = memo(SvgComponent);
export default Memo;"
`;

exports[`plugin typescript with "namedExport" and "exportType" option and without "previousExport" state exports via named export 1`] = `
"import * as React from \\"react\\";

const SvgComponent = () => <svg><g /></svg>;

export { SvgComponent as ReactComponent };"
`;

exports[`plugin typescript with "namedExport" option and "previousExport" state has custom named export 1`] = `
"import * as React from \\"react\\";

const SvgComponent = () => <svg><g /></svg>;

export { SvgComponent as Component };
var img = new Image();
img.src = '...';
export default img;"
`;

exports[`plugin typescript with "native" and "expandProps" option adds import from "react-native-svg" and adds props 1`] = `
"import * as React from \\"react\\";
import Svg, { SvgProps } from \\"react-native-svg\\";

const SvgComponent = (props: SvgProps) => <Svg><g /></Svg>;

export default SvgComponent;"
`;

exports[`plugin typescript with "native" option adds import from "react-native-svg" 1`] = `
"import * as React from \\"react\\";
import Svg from \\"react-native-svg\\";

const SvgComponent = () => <Svg><g /></Svg>;

export default SvgComponent;"
`;

exports[`plugin typescript with "native", "ref" and "expandProps" option adds import from "react-native-svg" and adds props and adds ForwardRef component 1`] = `
"import * as React from \\"react\\";
import Svg, { SvgProps } from \\"react-native-svg\\";
import { Ref, forwardRef } from \\"react\\";

const SvgComponent = (props: SvgProps, ref: Ref<SVGSVGElement>) => <Svg><g /></Svg>;

const ForwardRef = forwardRef(SvgComponent);
export default ForwardRef;"
`;

exports[`plugin typescript with "native", "ref" option adds import from "react-native-svg" and adds ForwardRef component 1`] = `
"import * as React from \\"react\\";
import Svg from \\"react-native-svg\\";
import { Ref, forwardRef } from \\"react\\";

const SvgComponent = (_, ref: Ref<SVGSVGElement>) => <Svg><g /></Svg>;

const ForwardRef = forwardRef(SvgComponent);
export default ForwardRef;"
`;

exports[`plugin typescript with "ref" and "expandProps" option expands props 1`] = `
"import * as React from \\"react\\";
import { SVGProps, Ref, forwardRef } from \\"react\\";

const SvgComponent = (props: SVGProps<SVGSVGElement>, ref: Ref<SVGSVGElement>) => <svg><g /></svg>;

const ForwardRef = forwardRef(SvgComponent);
export default ForwardRef;"
`;

exports[`plugin typescript with "ref" option adds ForwardRef component 1`] = `
"import * as React from \\"react\\";
import { Ref, forwardRef } from \\"react\\";

const SvgComponent = (_, ref: Ref<SVGSVGElement>) => <svg><g /></svg>;

const ForwardRef = forwardRef(SvgComponent);
export default ForwardRef;"
`;

exports[`plugin typescript with "titleProp" adds "titleProp" and "titleId" prop 1`] = `
"import * as React from \\"react\\";
interface SVGRProps {
  title?: string;
  titleId?: string;
}

const SvgComponent = ({
  title,
  titleId
}: SVGRProps) => <svg><g /></svg>;

export default SvgComponent;"
`;

exports[`plugin typescript with "titleProp" and "expandProps" adds "titleProp", "titleId" props and expands props 1`] = `
"import * as React from \\"react\\";
import { SVGProps } from \\"react\\";
interface SVGRProps {
  title?: string;
  titleId?: string;
}

const SvgComponent = ({
  title,
  titleId,
  ...props
}: SVGProps<SVGSVGElement> & SVGRProps) => <svg><g /></svg>;

export default SvgComponent;"
`;

exports[`plugin typescript with both "memo" and "ref" option wrap component in "React.memo" and "React.forwardRef" 1`] = `
"import * as React from \\"react\\";
import { Ref, forwardRef, memo } from \\"react\\";

const SvgComponent = (_, ref: Ref<SVGSVGElement>) => <svg><g /></svg>;

const ForwardRef = forwardRef(SvgComponent);
const Memo = memo(ForwardRef);
export default Memo;"
`;
